Fixes #7708: Improve appearance of the webhelp doc

amousset opened the pull request:
https://www.rudder-project.org/redmine/issues/7708

amousset commented:
Not totally finished, still missing:
* Add search field from DocSearch for public doc
* One page mode for embedded doc

amousset commented:
![doc](https://cloud.githubusercontent.com/assets/329388/12457598/b0a332aa-bfa5-11e5-9bd7-86a92cc13aaa.png)

jooooooon commented on template/common/jquery/treeview/jquery.treeview.css:
```diff
@@ -62,10 +62,10 @@
+/*.filetree span.folder { background: url(images/folder.gif) 0 0 no-repeat; }
 .filetree li.expandable span.folder { background: url(images/folder-closed.gif) 0 0 no-repeat; }
-.filetree span.file { background: url(images/file.gif) 0 0 no-repeat; }
+.filetree span.file { background: url(images/file.gif) 0 0 no-repeat; }*/
```
?

amousset commented:
![doc](https://cloud.githubusercontent.com/assets/329388/12477608/c88fea8a-c030-11e5-9127-f1b54f5a6178.png)

amousset commented:
PR updated

jooooooon commented:
I like the list of versions at top right. However, I think it could be clarified a little: could we put in bold (or some other way of highlighting) the version we're on currently? And maybe add a title "Version: " in front of the list, to clarify what we're talking about?

jooooooon commented on xsl/xsl-ns-stylesheets/xhtml/chunk-common.xsl:
```diff
@@ -1209,7 +1209,7 @@ xmlns:exsl="http://exslt.org/common" xmlns:cf="http://docbook.sourceforge.net/xm
       <xsl:call-template name="make.legalnotice.head.links"/>
     </xsl:if>
-
+<link rel="stylesheet" href="https://cdn.jsdelivr.net/docsearch.js/1/docsearch.min.css" />
```
This won't work on embedded versions of the documentation. Do we have an idea to work around this yet?

amousset replied:
Yes! There are now two targets in the Makefile, webhelp (with DocSearch, for public doc) and webhelp-search (which includes the build-in doc, used previously, for embedded doc, it also remove all calls to DocSearch from the doc).

jooooooon replied:
Ahhhhhhhhhhhhhh I see. Those names are very confusing to me, I had understood the exact opposite!

How about we rename webhelp-search webhelp-localsearch, to clarify?

jooooooon commented on xsl/webhelp-common.xsl:
```diff
@@ -525,6 +525,13 @@ border: none; background: none; font-weight: none; color: none; }
                 <xsl:call-template name="user.footer.navigation"/>
+                <script type="text/javascript" src="https://cdn.jsdelivr.net/docsearch.js/1/docsearch.min.js"></script> 
+<script type="text/javascript"> docsearch({
+apiKey: '339fdb9d06e694d14046ad9a5cd4fbb3',
```
Should this be public?

amousset replied:
I don't see how it could not, as requests are made by the client.

jooooooon replied:
True.

amousset commented:
PR updated

jooooooon commented on xsl/webhelp-common.xsl:
```diff
@@ -525,6 +525,15 @@ border: none; background: none; font-weight: none; color: none; }
                 </div>
 
                 <xsl:call-template name="user.footer.navigation"/>
+                <xsl:choose><xsl:when test="$webhelp.include.search.tab = '0'">
```
Shouldn't this be the opposite? That is, test if equal to 1?

I'm not big on XSLT so I may just be confused... but in the Makefile above, you set this variable to 1 to include the search box, and here you seem to be including it when it's 0...

jooooooon replied:
You can ignore this, I know see that this variable does the opposite of what I thought.

jooooooon commented:
Great idea, by the way :)

amousset commented:
PR updated

amousset commented:
PR updated

amousset commented:
![doc2](https://cloud.githubusercontent.com/assets/329388/12481575/643d8f8c-c04a-11e5-8ce9-0015f2f12aef.png)

jooooooon commented:
Looks good. But when I run it on my system, the results box of the search is cut off at the bottom of the orange box.

amousset commented:
Here too! @VinceMacBuche will help me on this, it goes beyond my web skills (but you can manually set the z-index of the #header div to 1 to test the search box).

amousset commented:
PR updated

amousset commented on xsl/webhelp-common.xsl:
```diff
@@ -639,9 +654,27 @@ border: none; background: none; font-weight: none; color: none; }
+      <span>Version: <strong>2.11 (ESR)</strong> | <a href="http://www.rudder-project.org/doc-3.0/">3.0</a> | 
+                                                   <a href="http://www.rudder-project.org/doc-3.1/">3.1</a> |
+                                                   <a href="http://www.rudder-project.org/doc-3.2/">3.2</a></span>
+      <span>Download as: <a href="/rudder-doc-2.11/rudder-doc.pdf">epub</a> | <a href="/rudder-doc-2.11/rudder-doc.pdf">pdf</a></span>
```
We have to be careful for these links because they won't work on the embedded version in 3.2.
We could:
* put pdf and epub in the webhelp folder for Rudder >= 3.2 (as webhelp becomes the default) and redirect /rudder-doc-3.2/rudder-doc.pdf to /doc-3.2/rudder-doc.pdf in public docs, to keep the same layout everywhere
* Have a special case to use different links for embedded docs

amousset replied:
We need a special case anyway as the epub is not embedded, we will need to use an external link.

amousset commented:
PR updated

VinceMacBuche commented:
Seen the search box display issue with @amousset , it was just a question of css application priority ... like always with css :dancer:

amousset commented:
PR updated

amousset commented:
:tada: Search box is now working!

VinceMacBuche commented:
:beers: for everyone!

amousset commented:
PR updated

amousset commented:
What is missing before merging:
* An doc index for each major version (I'm waiting for the answer)
* A decision about files layout (https://github.com/Normation/rudder-doc/pull/158/files#r50400719)
* Prepare folders for the localsearch doc on the server

What has to be done when merged:
* Change the build tasks to use the new folders (webhelp and webhelp-localsearch)

amousset commented:
PR updated

amousset commented:
PR updated

amousset commented:
Improved search box (and changed admonition icons).
![doc3](https://cloud.githubusercontent.com/assets/329388/12485261/4b3fe142-c05e-11e5-9d75-47292ba09df3.png)

amousset commented:
PR updated

amousset commented:
Docsearch now takes a version parameter, and all are docs are indexed :+1:

amousset commented:
PR updated

amousset commented:
PR updated

amousset commented:
Added a variable to trigger local link to pdf from the Makefile.

amousset commented:
PR ready!

amousset commented:
The localsearch docs should be added to robots.txt on the server.

jooooooon commented on rudder-doc.txt:
```diff
@@ -1,15 +1,15 @@
-Rudder User Documentation
+Rudder 2.11 - User Manual
```
This will require an additional PR on each branch to fix

jooooooon replied:
(that's not a problem, just a reminder)

amousset replied:
We also need to adjust DocSearch parameters to query the right index.

jooooooon commented on xsl/webhelp-common.xsl:
```diff
@@ -256,7 +256,7 @@ border: none; background: none; font-weight: none; color: none; }
 	.ui-layout-pane { /* all 'panes' */
 		background: #FFF;
-		border: 1px solid #BBB;
+		*/border: 1px solid #BBB;*/
```
Remove it if you don't want it. git can remember it for us, we don't need to carry around commented CSS

amousset replied:
I missed this one, removed.

amousset commented:
PR updated

jooooooon commented on Makefile:
```diff
@@ -85,18 +86,32 @@ $(LUCENE_ANALYZER_JAR):
 
-docs/index.html: $(SOURCES)
-	mkdir -p docs
+webhelp/index.html: $(SOURCES)
```
Why change this directory name? It will be very inconvient, as we'll have to change all the jobs in the CI to copy a different directory, etc... Could we not just keep the "main" name as is?

jooooooon replied:
I don't mind changing it if you have a good reason (and are willing to do the changes in Jenkins :p) I just think I might be missing the point.

amousset replied:
The reason it was `docs` is that I could not find the way to change it in first place, but I would have preferred a meaningful name.
I can put localsearch in `docs-localsearch` and keep the `webhelp` and `webhelp-localsearch` targets, what do you think is the best option?

amousset commented:
PR updated

amousset commented:
![resources](https://cloud.githubusercontent.com/assets/329388/12613710/a5855adc-c4fb-11e5-92db-0121a5247b41.png)

amousset commented:
Added a line to point to other resources.

jooooooon commented on xsl/webhelp-common.xsl:
```diff
@@ -639,9 +656,37 @@ border: none; background: none; font-weight: none; color: none; }
+    <xsl:template name="otherlinks">
+    <div id="otherlinks">
+      <span>Resources: <strong>User manual</strong> |
+                             <a href="http://www.rudder-project.org/foswiki/bin/view/System/Documentation:ChangeLog211">Changelog</a> | 
```
This is a great idea! However, we *know* that we want to change the URLs for these changelogs, so before merging this PR and sending these URLs out as static code to Rudder installations everywhere, we should address this.

I propose a simple approach: we make a memorable, simple URL on rudder-project.org/changelog-<version> (same format at /doc-<version> and same shortcuts such as /doc, /doc-latest, etc), and use Apache rewrite rules to make sure that URL always goes to the right page. That way, we can change the actual URL no problem, anytime.

I've added the rewrite rules already, can you adjust the PR?

amousset replied:
Done!

amousset commented:
PR updated

amousset commented:
Added a `$rudder.version` variable (but could not use it inside DocSearch javascript, so we will still have to change it manually).

amousset commented:
PR updated

amousset commented:
PR updated

jooooooon commented:
I've been looking at this on my machine, and I see a few items that I think still need work:

* I'm less keen on the sidebar presentation compared to the original. I miss the hover color on the links, and the spacing around the whole list in particular. I agree the icons weren't great, but it feels bare without them.
* Some, it not all, links within in a page don't seem to work. For me, webhelp/search-nodes.html#advanced-search doesn't scroll down to the h3 title in that page
* Some images are wider than the screen. In the old old docs, these were ajusted automatically to be 100% width (or something similar). This is lacking, because some of our screenshots are HUGE.
* Previously, the content was aligned on the left, next to the sidebar. Now it is centered, with grey space around the reading area. I'm not sure I like this - it makes the page look empty, and therefore the docs "feel" less complete. I know this is purely psychological, but I think I'd be happier if we returned to aligned on the left, and the whole right pane being white.

amousset commented:
PR updated

amousset commented:
* Sidebar: reverted to original appearance, will probably change the icons later though
* Image width is now limited

amousset commented:
PR updated

amousset commented:
* Link issue fixed with the help of @VinceMacBuche :+1:

amousset commented:
PR updated

amousset commented:
Switched to white background, put content on the left. @jooooooon What do you think of these changes?
![white](https://cloud.githubusercontent.com/assets/329388/12821992/20519d50-cb66-11e5-832b-b773b945a85c.png)

amousset commented:
PR updated

amousset commented:
Remove the gray line at the top of the left tree.

amousset commented:
PR updated

amousset commented:
PR updated

amousset commented:
PR updated

amousset commented:
Added the Lato font
![lato2](https://cloud.githubusercontent.com/assets/329388/12842285/1bc35706-cbf2-11e5-87f3-147fa75989a3.png)

VinceMacBuche commented:
So much awosemness

jooooooon commented:
A lot of nice improvements! Good work, @amousset.

There are still a few issues that need addressing here, in my opinion.

1) As I said before, comparing the sidebar as it was and this new one, I miss the hover color on the links. I really dislike the new icons, too (this is probably because when I started using a computer back in 1988, these are the exact icons that computer had!!).

2) There seems to be an overlap problem with titles that are too wide in the sidebar.:
![image](https://cloud.githubusercontent.com/assets/93237/12948802/b38925aa-d004-11e5-810d-2b02aa360cb6.png)

In fact, in general, I think the font size in the left sidebar is a bit big - hardly any titles do fit in the available width. How about reducing it?

3) We used to have Prev/Next links at the bottom of each section, but they're gone now. I think these are a really important navigation feature, for multi-page docs. This is *exactly* how I read such docs: read a section, get to the bottom, click Next. Now I can't do that on our doc. Can you put them back please?

amousset commented:
The overlap issue is strange, I do not see it in any browser, what is your browser/browser version/screen resolution?

On my Chromium for example, the text is wrapped:
![chromium](https://cloud.githubusercontent.com/assets/329388/12954370/8a07868a-d01e-11e5-842f-487d1cef4443.png)

jooooooon commented:
It's Version 44.0.2403.89 Built on 8.1, running on Debian 8.2 (64-bit). Screen resolution is 1920x1080.

amousset commented:
PR updated

amousset commented:
* Added back the links in the footer (I do not clearly remember why I removed them...)
* Added an orange color on hover for toc items

amousset commented:
Chromium 47 here, I'm looking for older version packages.

amousset commented:
PR updated

amousset commented:
Changed the sidebar icons, hope they do not look like they come from the 80s this time. Also styled the footer links, and added a very light gray background to the left tree.
![footer](https://cloud.githubusercontent.com/assets/329388/12973030/26a44a88-d0aa-11e5-813b-8fd6bc62f3d8.png)
![icon](https://cloud.githubusercontent.com/assets/329388/12973031/29112a84-d0aa-11e5-86ab-cd50cb06949b.png)

amousset commented:
PR updated

amousset commented:
I could reproduce your overlapping issue with Chromium 44, it seems to be a bug that has been fixed since (works in Chromium 47). Added a workaround with the help of @VinceMacBuche :+1:

VinceMacBuche commented:
Web is so broken ...

jooooooon commented:
Yes, indeed it is fixed! Nice work.

I like the side bar and icons a lot better now, thanks :)

jooooooon commented:
I no longer see any issues on my machine, nor any major missing parts (this is a scientific test that takes it from "works for you" to "works for you and me", thus ensuring a much higher percentage of expected reliability in production :) )

Therefore I'm going to merge this. Any remaining issues can be fixed with extra pull requests as needed.

jooooooon commented:
Please provide pull requests for the required version updates on 3.0/3.1/3.2/3.3 branches
